src/Lexer.ts

```diff
@@ -0,0 +1,98 @@
+function* Lexer(input: string) {
+  const inputAsArr: arrayOfString = Array.from(input);
+  inputAsArr.push(TokenType.eof.toString());
+
+  let position: number = 0;
+  for (; position < inputAsArr.length; position++) {
+    const str: string = inputAsArr[position];
+    const { token, updatedPos } = identifyLiteral({ str, position, inputAsArr });
+    updatedPos && (position = updatedPos);
+
+    yield token;
+  }
+}
+
+// identifyLiteral helps to identify literals i.e either the token, number or identifier
+type identifyLiteralParams = {
+  str: string;
+  position: number;
+  inputAsArr: arrayOfString;
+};
+const identifyLiteral = (data: identifyLiteralParams): identifyLiteralReturnType => {
+  const { str, position, inputAsArr } = data;
+  let token, updatedPos;
+
+  switch (str) {
+    case '.':
+      token = { type: TokenType.dot, literal: str };
+      break;
+    case '[':
+      token = { type: TokenType.lbracket, literal: str };
+      break;
+    case ']':
+      token = { type: TokenType.rbracket, literal: str };
+      break;
+    default:
+      if (isNumeric(str)) {
+        const { updatedPosInt, int } = readNumber(position, inputAsArr) || {};
+        updatedPosInt && (updatedPos = updatedPosInt);
+        token = { type: TokenType.int, literal: int };
+      } else {
+        const { updatedPosIdent, ident } = readIdentifier(position, inputAsArr) || {};
+        updatedPosIdent && (updatedPos = updatedPosIdent);
+        token = { type: TokenType.ident, literal: ident };
+      }
+  }
+
+  return { token, updatedPos };
+};
+
+// isNumeric helps identify if the given string is a number or not
+const isNumeric = (n: string): boolean => {
+  const num: number = parseFloat(n);
+
+  return !isNaN(num) && isFinite(num);
+};
+
+// readNumber reads the integer next to the token
+type readNumberReturnType = { int: string; updatedPosInt: number } | undefined;
+const readNumber = (pos: number, srcInput: arrayOfString): readNumberReturnType => {
+  let int: string = '';
+  const prevToken: string = srcInput[pos - 1];
+  for (; pos < srcInput.length; pos++) {
+    if (prevToken === '[' && srcInput[pos] === ']') {
+      return { int, updatedPosInt: pos - 1 };
+    }
+    int += srcInput[pos];
+  }
+};
+
+// readIdentifier reads the identfier next to the token
+type readIdentifierReturnType = { ident: string; updatedPosIdent: number } | undefined;
+const readIdentifier = (pos: number, srcInput: arrayOfString): readIdentifierReturnType => {
+  let ident: string = '';
+  for (; pos < srcInput.length; pos++) {
+    if (!isToken(srcInput[pos])) {
+      return { ident, updatedPosIdent: pos - 1 };
+    }
+    if (srcInput[pos] === TokenType.eof.toString()) {
+      return { ident, updatedPosIdent: srcInput.length };
+    }
+    ident += srcInput[pos];
+  }
+};
+
+// isToken finds if the given character is in the list of tokens
+const isToken = (ch: string): boolean => ch != '.' && ch != '[' && ch != ']';
```

Comment:
Don't you mean `isIdent`? Because you're checking here if the input is not any of the tokens, which means it's an identifier. Why even rename it from the reference implementation? What a cognitive overhead to introduce when the reference is already working in production.